src/etos_test_runner/lib/testrunner.py

```diff
@@ -14,17 +14,75 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """ETR test runner module."""
+import json
 import time
 import os
 import logging
 from pprint import pprint
+from typing import Union
 
 from etos_test_runner.lib.iut_monitoring import IutMonitoring
 from etos_test_runner.lib.executor import Executor
 from etos_test_runner.lib.workspace import Workspace
 from etos_test_runner.lib.log_area import LogArea
 
 
+class CustomVerdictMatcher:
+    # pylint: disable=too-few-public-methods
+    """Match testframework output against user-defined verdict rules.
+
+    Example rule definition:
+
+    rules = [
+        {
+            "description": "Test collection error, no artifacts created",
+            "condition": {
+                "test_framework_exit_code": 4,
+            },
+            "conclusion": "FAILED",
+            "verdict": "FAILED",
+        }
+    ]
+    """
+
+    SUPPORTED_CONDITION_KEYWORDS = [
+        "test_framework_exit_code",
+    ]
+
+    def __init__(self, rules: list, test_framework_output: dict) -> None:
+        """Create new instance."""
+        self.rules = rules
+        self.test_framework_output = test_framework_output
+
+        for rule in self.rules:
+            for key in rule["condition"].keys():
+                if key not in self.SUPPORTED_CONDITION_KEYWORDS:
+                    raise ValueError(
+                        f"Unsupported condition keyword for test outcome rules: {key}! "
+                        f"Supported keywords: {self.SUPPORTED_CONDITION_KEYWORDS}."
+                    )
```

Comment:
This error would be happening quite late in the execution. We should fail fast by initializing the `CustomVerdictMatcher` before running tests and instead of passing `test_framework_output` in the constructor, we should pass it to `evaluate`

Reply:
Done